File: example-jv-04retrofit/src/androidTest/java/foo/bar/example/foreretrofit/ui/fruit/FruitViewTestStateBuilder.java
package foo.bar.example.foreretrofit.ui.fruit;

import androidx.test.InstrumentationRegistry;
import androidx.test.rule.ActivityTestRule;
import co.early.fore.core.WorkMode;
import foo.bar.example.foreretrofit.App;
import foo.bar.example.foreretrofit.OG;
import foo.bar.example.foreretrofit.ProgressBarIdler;
import foo.bar.example.foreretrofit.api.fruits.FruitPojo;
import foo.bar.example.foreretrofit.feature.fruit.FruitFetcher;

import static org.mockito.Mockito.when;

/**
 *
 */
public class FruitViewTestStateBuilder {

    private FruitFetcher mockFruitFetcher;

    FruitViewTestStateBuilder(FruitFetcher mockFruitFetcher) {
        this.mockFruitFetcher = mockFruitFetcher;
    }

    FruitViewTestStateBuilder isBusy(boolean busy) {
        when(mockFruitFetcher.isBusy()).thenReturn(busy);
        return this;
    }

    FruitViewTestStateBuilder hasFruit(FruitPojo fruitPojo) {
        when(mockFruitFetcher.getCurrentFruit()).thenReturn(fruitPojo);
        return this;
    }

    ActivityTestRule<FruitActivity>  createRule(){

        return new ActivityTestRule<FruitActivity>(FruitActivity.class) {
            @Override
            protected void beforeActivityLaunched() {

                //get hold of the application
                App app = (App) InstrumentationRegistry.getTargetContext().getApplicationContext();
                OG.setApplication(app, WorkMode.SYNCHRONOUS);

                //inject our mocks so our UI layer will pick them up
                OG.putMock(FruitFetcher.class, mockFruitFetcher);

                app.registerActivityLifecycleCallbacks(new ProgressBarIdler());
            }

        };
    }

}

File: example04retrofit-kt/src/androidTest/java/foo/bar/example/foreretrofit/ui/fruit/FruitViewTestStateBuilder.kt

package foo.bar.example.foreretrofit.ui.fruit

import androidx.test.InstrumentationRegistry
import androidx.test.rule.ActivityTestRule
import co.early.fore.core.WorkMode
import foo.bar.example.foreretrofit.ProgressBarIdler
import foo.bar.example.foreretrofitcoroutine.App
import foo.bar.example.foreretrofitcoroutine.OG
import foo.bar.example.foreretrofitcoroutine.api.fruits.FruitPojo
import foo.bar.example.foreretrofitcoroutine.feature.fruit.FruitFetcher
import foo.bar.example.foreretrofitcoroutine.ui.fruit.FruitActivity
import io.mockk.every


class FruitViewTestStateBuilder internal constructor(private val mockFruitFetcher: FruitFetcher) {

    internal fun isBusy(busy: Boolean): FruitViewTestStateBuilder {
        every { mockFruitFetcher.isBusy } returns (busy)
        return this
    }

    internal fun hasFruit(fruitPojo: FruitPojo): FruitViewTestStateBuilder {
        every { mockFruitFetcher.currentFruit } returns (fruitPojo)
        return this
    }

    internal fun createRule(): ActivityTestRule<FruitActivity> {

        return object : ActivityTestRule<FruitActivity>(FruitActivity::class.java) {
            override fun beforeActivityLaunched() {

                //get hold of the application
                val application = InstrumentationRegistry.getTargetContext().applicationContext as App
                application.registerActivityLifecycleCallbacks(ProgressBarIdler())

                //inject our mocks so our UI layer will pick them up
                OG.setApplication(application, WorkMode.SYNCHRONOUS)
                OG.putMock(FruitFetcher::class.java, mockFruitFetcher)
            }
        }
    }

}
